### DISCUSSION TITLE
STJ: Preserving comparers with Immutable Dictionary/Set

### DISCUSSION THREAD

**Discussion by TonyValenti** [dotnet/runtime#55243] [Ideas]
Would it be possible to have the immutable collection deserializers enhanced so that they use the same Comparer of the original member?

(Please excuse typos. I'm on my phone)

For example:

Public Record Foo {
 public ImmutableSet<string> Member {get; init;} = ImmutableSet<string>.Empty.WithComparer(StringComparer.InvariantCultureIgnoreCase);
}

When member is deserialized, preserve InvariantCultureIgnoreCase. 

Or, alternatively, instead of creating a new ImmutableSet and replacing the existing one, just call Union on the existing one and assign to that value (so it would inherit initial values too).

**Comment by Joe4evr** [comment]
The problem here is what are you going to do if one of these collections is instantiated with `MyThirdPartyComparer`?

I've actually had an idea around this for a while, and I'm curious if there's any merit: It starts with having some member on the built-in (concrete) `StringComparer` types that would produce a magic string that encapsulates that particular comparer's configuration (ie. ordinal/culture/invariant + case-[in]sensitve), which could then be serialized alongside the collection. On deserialization, this string could be passed into another API which gets the appropriate comperer to use, eg.
```diff
public abstract partial class StringComparer
{
     // Returns a known StringComparer based on the magic string, or null if the string didn't match
+    public static StringComparer? GetWellKnownComparer(string description);
}
```

For serialization, things are a bit tricky.
1. There should be a way for a relevant collection to expose its comparer.
    * Some of those in the BCL already do, but not all of them.
    * Third-party collections would want to opt-in to this mechanism as well.
2. This entire behavior should probably be opt-in anyway.

Number 2 could probably be done with an attribute to instruct the (de)serializer:
```cs
// This attribute doesn't exist, but take it as a strawman
[JsonCollection(IncludeComparerIfAble = true)]
public HashSet<string> Strings { get; init; }

// Has no effect, since it's entirely special-cased aroung String comparers (initially)
[JsonCollection(IncludeComparerIfAble = true)]
public HashSet<int> Ints { get; init; }
```

Number 1 could be solved by way of a new interface:
```cs
namespace System.Collections.Generic
{
    public interface IKeyedCollection<in TKey> // Open question: extend 'IEnumerable<T>'?
    {
        IEqualityComparer<TKey> Comparer { get; }
    }
}
```
This would allow the relevant converter to do the following:
```cs
// serialize
if (_includeComparer && collection is IKeyedCollection<string> { Comparer: StringComparer comparer })
{
    // Write out the magic string as a json member
}

// deserialize
IEqualityComparer<TKey>? comparer;
if (_includeComparer && isStringKeyedCollection)
{
    string? comparerDesc = /* Read the magic string */;
    comparer = StringComparer.GetWellKnownComparer(comparerDesc);
}
```

There's probably a number of holes in this idea, but I'm curious if those aren't insurmountable.

### Notes:
* Limiting support to only the versions of `StringComparer` known by the BCL should avoid many problems known from Binary Serialization.
* The magic descriptor string should be well-structured so that [it isn't obfuscating any details about its meaning](https://en.wikipedia.org/wiki/Nothing-up-my-sleeve_number).
* Should the method to obtain the magic descriptor string be `public`? It would help to allow third-party code to utilize the same mechanic to de-/serialize a `StringComparer`, but I'm unsure about what shape that would have.
